File: facebook-core/src/main/java/com/facebook/appevents/iap/InAppPurchaseLoggerManager.java
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.appevents.iap;

import android.content.Context;
import android.content.SharedPreferences;
import androidx.annotation.RestrictTo;
import com.facebook.FacebookSdk;
import com.facebook.appevents.internal.AutomaticAnalyticsLogger;
import com.facebook.internal.instrument.crashshield.AutoHandleExceptions;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArraySet;
import org.json.JSONObject;

@AutoHandleExceptions
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class InAppPurchaseLoggerManager {

  private static SharedPreferences sharedPreferences;
  private static final Set<String> cachedPurchaseSet = new CopyOnWriteArraySet<>();
  private static final Map<String, Long> cachedPurchaseMap = new ConcurrentHashMap<>();

  private static final String PURCHASE_TIME = "purchaseTime";
  private static final String PRODUCT_DETAILS_STORE = "com.facebook.internal.iap.PRODUCT_DETAILS";
  private static final String LAST_CLEARED_TIME = "LAST_CLEARED_TIME";
  private static final String PURCHASE_DETAILS_SET = "PURCHASE_DETAILS_SET";
  private static final String LAST_QUERY_PURCHASE_HISTORY_TIME = "LAST_QUERY_PURCHASE_HISTORY_TIME";

  private static final int CACHE_CLEAR_TIME_LIMIT_SEC = 7 * 24 * 60 * 60; // 7 days
  private static final int PURCHASE_IN_CACHE_INTERVAL = 24 * 60 * 60; // 1 day

  private static void readPurchaseCache() {
    // clear cached purchases logged by lib 1
    SharedPreferences cachedSkuSharedPref =
        FacebookSdk.getApplicationContext()
            .getSharedPreferences("com.facebook.internal.SKU_DETAILS", Context.MODE_PRIVATE);
    SharedPreferences cachedPurchaseSharedPref =
        FacebookSdk.getApplicationContext()
            .getSharedPreferences("com.facebook.internal.PURCHASE", Context.MODE_PRIVATE);
    if (cachedSkuSharedPref.contains("LAST_CLEARED_TIME")) {
      cachedSkuSharedPref.edit().clear().apply();
      cachedPurchaseSharedPref.edit().clear().apply();
    }

    sharedPreferences =
        FacebookSdk.getApplicationContext()
            .getSharedPreferences(PRODUCT_DETAILS_STORE, Context.MODE_PRIVATE);
    cachedPurchaseSet.addAll(
        sharedPreferences.getStringSet(PURCHASE_DETAILS_SET, new HashSet<String>()));

    // Construct purchase de-dup map.
    for (String purchaseHistory : cachedPurchaseSet) {
      String[] splitPurchase = purchaseHistory.split(";", 2);
      cachedPurchaseMap.put(splitPurchase[0], Long.parseLong(splitPurchase[1]));
    }

    // Clean up cache every 7 days, and only keep recent 1 day purchases
    clearOutdatedProductInfoInCache();
  }

  public static void filterPurchaseLogging(
      Map<String, JSONObject> purchaseDetailsMap, Map<String, JSONObject> skuDetailsMap) {
    readPurchaseCache();

    Map<String, String> loggingReadyMap =
        new HashMap<>(
            constructLoggingReadyMap(cacheDeDupPurchase(purchaseDetailsMap), skuDetailsMap));
    logPurchases(loggingReadyMap);
  }

  static void logPurchases(Map<String, String> purchaseDetailsMap) {
    for (Map.Entry<String, String> pair : purchaseDetailsMap.entrySet()) {
      String purchaseDetails = pair.getKey();
      String skuDetails = pair.getValue();
      if (purchaseDetails != null && skuDetails != null) {
        AutomaticAnalyticsLogger.logPurchase(purchaseDetails, skuDetails, false);
      }
    }
  }

  static Map<String, JSONObject> cacheDeDupPurchase(Map<String, JSONObject> purchaseDetailsMap) {
    long nowSec = System.currentTimeMillis() / 1000L;
    Map<String, JSONObject> tempPurchaseDetailsMap = new HashMap<>(purchaseDetailsMap);
    for (Map.Entry<String, JSONObject> entry : tempPurchaseDetailsMap.entrySet()) {
      try {
        JSONObject purchaseJson = entry.getValue();
        if (purchaseJson.has("purchaseToken")) {
          String purchaseToken = purchaseJson.getString("purchaseToken");
          if (cachedPurchaseMap.containsKey(purchaseToken)) {
            purchaseDetailsMap.remove(entry.getKey());
          } else {
            cachedPurchaseSet.add(purchaseToken + ';' + nowSec);
          }
        }
      } catch (Exception e) {
        /* swallow */
      }
    }

    sharedPreferences.edit().putStringSet(PURCHASE_DETAILS_SET, cachedPurchaseSet).apply();
    return new HashMap<>(purchaseDetailsMap);
  }

  private static void clearOutdatedProductInfoInCache() {
    long nowSec = System.currentTimeMillis() / 1000L;
    long lastClearedTimeSec = sharedPreferences.getLong(LAST_CLEARED_TIME, 0);

    if (lastClearedTimeSec == 0) {
      sharedPreferences.edit().putLong(LAST_CLEARED_TIME, nowSec).apply();
    } else if ((nowSec - lastClearedTimeSec) > CACHE_CLEAR_TIME_LIMIT_SEC) {
      Map<String, Long> tempPurchaseMap = new HashMap<>(cachedPurchaseMap);
      for (Map.Entry<String, Long> historyPurchase : tempPurchaseMap.entrySet()) {
        String purchaseToken = historyPurchase.getKey();
        Long historyPurchaseTime = historyPurchase.getValue();
        if (nowSec - historyPurchaseTime > PURCHASE_IN_CACHE_INTERVAL) {
          cachedPurchaseSet.remove(purchaseToken + ";" + historyPurchaseTime);
          cachedPurchaseMap.remove(purchaseToken);
        }
      }
      sharedPreferences
          .edit()
          .putStringSet(PURCHASE_DETAILS_SET, cachedPurchaseSet)
          .putLong(LAST_CLEARED_TIME, nowSec)
          .apply();
    }
  }

  public static boolean eligibleQueryPurchaseHistory() {
    readPurchaseCache();
    long nowSec = System.currentTimeMillis() / 1000L;
    long lastQueryPurchaseHistoryTime =
        sharedPreferences.getLong(LAST_QUERY_PURCHASE_HISTORY_TIME, 0);
    if (lastQueryPurchaseHistoryTime != 0
        && nowSec - lastQueryPurchaseHistoryTime < PURCHASE_IN_CACHE_INTERVAL) {
      return false;
    }
    sharedPreferences.edit().putLong(LAST_QUERY_PURCHASE_HISTORY_TIME, nowSec).apply();
    return true;
  }

  static Map<String, String> constructLoggingReadyMap(
      Map<String, JSONObject> purchaseDetailsMap, Map<String, JSONObject> skuDetailsMap) {
    long nowSec = System.currentTimeMillis() / 1000L;
    Map<String, String> purchaseResultMap = new HashMap<>();
    for (Map.Entry<String, JSONObject> pair : purchaseDetailsMap.entrySet()) {
      JSONObject skuDetail = skuDetailsMap.get(pair.getKey());
      JSONObject purchaseDetail = pair.getValue();
      if (purchaseDetail != null && purchaseDetail.has(PURCHASE_TIME)) {
        try {
          long purchaseTime = purchaseDetail.getLong(PURCHASE_TIME);
          // Purchase is too old (more than 24h) to log
          if (nowSec - purchaseTime / 1000L > PURCHASE_IN_CACHE_INTERVAL) {
            continue;
          }
          if (skuDetail != null) {
            purchaseResultMap.put(purchaseDetail.toString(), skuDetail.toString());
          }
        } catch (Exception e) {
          /* swallow */
        }
      }
    }
    return purchaseResultMap;
  }
}

File: facebook-core/src/main/java/com/facebook/appevents/iap/InAppPurchaseLoggerManager.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.appevents.iap

import android.content.Context
import androidx.annotation.RestrictTo
import androidx.annotation.VisibleForTesting
import com.facebook.FacebookSdk.getApplicationContext
import com.facebook.appevents.internal.AutomaticAnalyticsLogger.logPurchase
import com.facebook.appevents.internal.Constants
import com.facebook.internal.instrument.crashshield.AutoHandleExceptions
import java.util.HashMap
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArraySet
import org.json.JSONObject
import kotlin.Exception
import kotlin.math.abs
import kotlin.math.log10
import kotlin.math.max
import kotlin.math.min

@AutoHandleExceptions
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
object InAppPurchaseLoggerManager {
    // January 10, 2025 05:00:00 PM GMT
    private const val APPROXIMATE_IAP_ENHANCEMENT_RELEASE_TIME = 1736528400000L
    private const val MILLISECONDS_IN_SECONDS = 1000.0
    private const val PURCHASE_TIME = "purchaseTime"
    private const val IAP_SKU_CACHE_GPBLV1 = "com.facebook.internal.SKU_DETAILS"
    private const val IAP_PURCHASE_CACHE_GPBLV1 = "com.facebook.internal.PURCHASE"
    private const val IAP_CACHE_OLD = "com.facebook.internal.iap.PRODUCT_DETAILS"

    private const val IAP_CACHE_GPBLV2V7 = "com.facebook.internal.iap.IAP_CACHE_GPBLV2V7"
    private const val CACHED_PURCHASES_KEY = "PURCHASE_DETAILS_SET"
    private const val APP_HAS_BEEN_LAUNCHED_KEY = "APP_HAS_BEEN_LAUNCHED_KEY"
    private const val TIME_OF_LAST_LOGGED_PURCHASE_KEY = "TIME_OF_LAST_LOGGED_PURCHASE"
    private const val TIME_OF_LAST_LOGGED_SUBSCRIPTION_KEY = "TIME_OF_LAST_LOGGED_SUBSCRIPTION"

    @JvmStatic
    fun migrateOldCacheHistory() {
        val iapCache =
            getApplicationContext().getSharedPreferences(
                IAP_CACHE_GPBLV2V7,
                Context.MODE_PRIVATE
            )
        var newestCandidateTime = max(
            max(
                iapCache.getLong(
                    TIME_OF_LAST_LOGGED_PURCHASE_KEY,
                    0L
                ),
                iapCache.getLong(
                    TIME_OF_LAST_LOGGED_SUBSCRIPTION_KEY,
                    0L
                )
            ), APPROXIMATE_IAP_ENHANCEMENT_RELEASE_TIME
        )
        val cachedPurchaseSet: MutableSet<String> = CopyOnWriteArraySet()
        val sharedPreferences =
            getApplicationContext().getSharedPreferences(
                IAP_CACHE_OLD,
                Context.MODE_PRIVATE
            )
        if (sharedPreferences.contains(CACHED_PURCHASES_KEY)) {
            cachedPurchaseSet.addAll(
                sharedPreferences.getStringSet(CACHED_PURCHASES_KEY, hashSetOf()) ?: hashSetOf()
            )

            // Construct purchase de-dup map.
            for (purchaseHistory in cachedPurchaseSet) {
                try {
                    val splitPurchase = purchaseHistory.split(";", limit = 2)
                    val timeInMilliseconds = splitPurchase[1].toLong() * 1000L

                    // Don't migrate any invalid values in cache
                    val digitsInTime = timeInMilliseconds.toString().length
                    val digitsInReleaseTime =
                        APPROXIMATE_IAP_ENHANCEMENT_RELEASE_TIME.toString().length
                    if (abs(digitsInTime - digitsInReleaseTime) >= log10(
                            MILLISECONDS_IN_SECONDS
                        )
                    ) {
                        continue
                    }
                    newestCandidateTime = max(newestCandidateTime, timeInMilliseconds)

                } catch (e: Exception) {
                    /* Swallow */
                    continue
                }
            }
        }
        iapCache.edit()
            .putLong(TIME_OF_LAST_LOGGED_SUBSCRIPTION_KEY, newestCandidateTime)
            .apply()
        iapCache.edit()
            .putLong(TIME_OF_LAST_LOGGED_PURCHASE_KEY, newestCandidateTime)
            .apply()
        deleteOldCacheHistory()
    }

    @JvmStatic
    fun deleteOldCacheHistory() {
        // clear cached purchases logged by lib 1
        val cachedSkuSharedPref =
            getApplicationContext()
                .getSharedPreferences(IAP_SKU_CACHE_GPBLV1, Context.MODE_PRIVATE)
        val cachedPurchaseSharedPref =
            getApplicationContext()
                .getSharedPreferences(IAP_PURCHASE_CACHE_GPBLV1, Context.MODE_PRIVATE)
        cachedSkuSharedPref.edit().clear().apply()
        cachedPurchaseSharedPref.edit().clear().apply()

        // clear cached purchases logged by lib 2 - 3
        val sharedPreferences =
            getApplicationContext().getSharedPreferences(
                IAP_CACHE_OLD,
                Context.MODE_PRIVATE
            )
        sharedPreferences.edit().clear().apply()
    }

    @JvmStatic
    fun getIsFirstAppLaunchWithNewIAP(): Boolean {
        val iapCache =
            getApplicationContext().getSharedPreferences(
                IAP_CACHE_GPBLV2V7,
                Context.MODE_PRIVATE
            )
        return !iapCache.contains(APP_HAS_BEEN_LAUNCHED_KEY)
    }

    @JvmStatic
    fun setAppHasBeenLaunchedWithNewIAP() {
        val iapCache =
            getApplicationContext().getSharedPreferences(
                IAP_CACHE_GPBLV2V7,
                Context.MODE_PRIVATE
            )
        try {
            iapCache.edit()
                .putBoolean(APP_HAS_BEEN_LAUNCHED_KEY, true)
                .apply()
        } catch (e: Exception) {
            /* Swallow */
        }
    }

    @JvmStatic
    fun updateLatestPossiblePurchaseTime() {
        try {
            val iapCache =
                getApplicationContext().getSharedPreferences(
                    IAP_CACHE_GPBLV2V7,
                    Context.MODE_PRIVATE
                )
            val currTime = System.currentTimeMillis()
            iapCache.edit()
                .putLong(TIME_OF_LAST_LOGGED_SUBSCRIPTION_KEY, currTime)
                .apply()
            iapCache.edit()
                .putLong(TIME_OF_LAST_LOGGED_PURCHASE_KEY, currTime)
                .apply()
        } catch (e: Exception) {
            /* Swallow */
        }
    }

    @JvmStatic
    fun filterPurchaseLogging(
        purchaseDetailsMap: MutableMap<String, JSONObject>,
        skuDetailsMap: Map<String, JSONObject?>,
        isSubscription: Boolean,
        packageName: String,
        billingClientVersion: InAppPurchaseUtils.BillingClientVersion,
        isFirstAppLaunch: Boolean,
    ) {
        val deduped = cacheDeDupPurchase(purchaseDetailsMap, isSubscription)
        val loggingReady = constructLoggingReadyMap(
            deduped,
            skuDetailsMap,
            packageName
        )
        logPurchases(loggingReady, isSubscription, billingClientVersion, isFirstAppLaunch)
    }

    private fun logPurchases(
        purchaseDetailsMap: Map<String, String>,
        isSubscription: Boolean,
        billingClientVersion: InAppPurchaseUtils.BillingClientVersion,
        isFirstAppLaunch: Boolean
    ) {
        for ((purchaseDetails, skuDetails) in purchaseDetailsMap) {
            logPurchase(
                purchaseDetails,
                skuDetails,
                isSubscription,
                billingClientVersion,
                isFirstAppLaunch
            )
        }
    }

    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    internal fun cacheDeDupPurchase(
        purchaseDetailsMap: MutableMap<String, JSONObject>,
        isSubscription: Boolean,
    ): Map<String, JSONObject> {
        val iapCache =
            getApplicationContext().getSharedPreferences(
                IAP_CACHE_GPBLV2V7,
                Context.MODE_PRIVATE
            )
        var timeOfLatestNewlyLoggedPurchase: Long = 0
        val timeOfLastLoggedPurchase: Long =
            if (isSubscription) {
                iapCache.getLong(
                    TIME_OF_LAST_LOGGED_SUBSCRIPTION_KEY,
                    APPROXIMATE_IAP_ENHANCEMENT_RELEASE_TIME
                )
            } else {
                iapCache.getLong(
                    TIME_OF_LAST_LOGGED_PURCHASE_KEY,
                    APPROXIMATE_IAP_ENHANCEMENT_RELEASE_TIME
                )
            }

        val tempPurchaseDetailsMap: Map<String, JSONObject> = purchaseDetailsMap.toMap()
        for ((key, purchaseJson) in tempPurchaseDetailsMap) {
            try {
                if (purchaseJson.has(Constants.GP_IAP_PURCHASE_TOKEN) && purchaseJson.has(
                        Constants.GP_IAP_PURCHASE_TIME
                    )
                ) {
                    val purchaseTime = purchaseJson.getLong(PURCHASE_TIME)
                    if (purchaseTime <= timeOfLastLoggedPurchase) {
                        purchaseDetailsMap.remove(key)
                    }

                    timeOfLatestNewlyLoggedPurchase =
                        max(timeOfLatestNewlyLoggedPurchase, purchaseTime)
                }
            } catch (e: Exception) {
                /* swallow */
            }
        }
        if (timeOfLatestNewlyLoggedPurchase >= timeOfLastLoggedPurchase) {
            if (isSubscription) {
                iapCache.edit()
                    .putLong(TIME_OF_LAST_LOGGED_SUBSCRIPTION_KEY, timeOfLatestNewlyLoggedPurchase)
                    .apply()
            } else {
                iapCache.edit()
                    .putLong(TIME_OF_LAST_LOGGED_PURCHASE_KEY, timeOfLatestNewlyLoggedPurchase)
                    .apply()
            }
        }
        return HashMap(purchaseDetailsMap)
    }

    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    internal fun constructLoggingReadyMap(
        purchaseDetailsMap: Map<String, JSONObject>,
        skuDetailsMap: Map<String, JSONObject?>,
        packageName: String
    ): Map<String, String> {
        val purchaseResultMap: MutableMap<String, String> = mutableMapOf()
        for ((key, purchaseDetail) in purchaseDetailsMap) {
            val skuDetail = skuDetailsMap[key]
            try {
                // Used during server-side processing of purchase verification
                purchaseDetail.put(InAppPurchaseConstants.PACKAGE_NAME, packageName)

                if (skuDetail != null) {
                    purchaseResultMap[purchaseDetail.toString()] = skuDetail.toString()
                }
            } catch (e: Exception) {
                /* swallow */
            }
        }
        return purchaseResultMap
    }
}
